File: ReverseWordsinaString2.py
import re

class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        def reverse(s):
            res = ''
            for i in range(len(s)-1, -1, -1):
                res += s[i]
            return res
            
        s = reverse(s)
        
        res = ''
        
        i = 0
        storeIndex = 0
        while i<len(s):
            if s[i] == ' ':
                i += 1
                continue
            j = i
            while j<len(s) and s[j]!=' ':
                j += 1
            res += reverse(s[i:j]) + ' '
            i = j
        
        return res.rstrip()
        
if __name__=='__main__':
    s = raw_input()
    solution = Solution()
    print(solution.reverseWords(s))

File: ReverseWordsinaString2.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    void reverseWords(string &s) {
        
        reverse(s, 0, s.length()-1);
        
        int storeIndex = 0;
        for(int i=0;i<s.length();i++){
            if(s[i]!=' '){
                if(storeIndex!=0) s[storeIndex++] = ' ';
                int j = i;
                while(j<s.length()&&s[j]!=' '){
                    s[storeIndex++] = s[j++];
                }
                reverse(s, storeIndex-(j-i), storeIndex-1);
                i = j;
            }
        }
        
        s.erase(s.begin()+storeIndex, s.end());
        
    }
    
    void reverse(string &s, int begin, int end){
        while(begin<end){
            char temp = s[begin];
            s[begin] = s[end];
            s[end] = temp;
            begin++;
            end--;
        }
    }
};

int main(){
    string s;
    getline(cin, s);
    
    Solution *solution = new Solution();
    solution->reverseWords(s);
    
    cout<<s;
    
    return 0;
}
